File: bindings/lua/proven/safe_currency.lua
-- SPDX-License-Identifier: AGPL-3.0-or-later
-- SPDX-FileCopyrightText: 2025 Hyperpolymath

--- Safe currency operations with type-safe monetary values.
--- @module proven.safe_currency

local safe_currency = {}

--- ISO 4217 currency codes with metadata.
safe_currency.currency_codes = {
    USD = { code = "USD", name = "US Dollar", symbol = "$", decimals = 2 },
    EUR = { code = "EUR", name = "Euro", symbol = "€", decimals = 2 },
    GBP = { code = "GBP", name = "British Pound", symbol = "£", decimals = 2 },
    JPY = { code = "JPY", name = "Japanese Yen", symbol = "¥", decimals = 0 },
    CHF = { code = "CHF", name = "Swiss Franc", symbol = "Fr", decimals = 2 },
    CAD = { code = "CAD", name = "Canadian Dollar", symbol = "$", decimals = 2 },
    AUD = { code = "AUD", name = "Australian Dollar", symbol = "$", decimals = 2 },
    NZD = { code = "NZD", name = "New Zealand Dollar", symbol = "$", decimals = 2 },
    CNY = { code = "CNY", name = "Chinese Yuan", symbol = "¥", decimals = 2 },
    INR = { code = "INR", name = "Indian Rupee", symbol = "₹", decimals = 2 },
    BRL = { code = "BRL", name = "Brazilian Real", symbol = "R$", decimals = 2 },
    MXN = { code = "MXN", name = "Mexican Peso", symbol = "$", decimals = 2 },
    KRW = { code = "KRW", name = "South Korean Won", symbol = "₩", decimals = 0 },
    SGD = { code = "SGD", name = "Singapore Dollar", symbol = "$", decimals = 2 },
    HKD = { code = "HKD", name = "Hong Kong Dollar", symbol = "$", decimals = 2 },
    SEK = { code = "SEK", name = "Swedish Krona", symbol = "kr", decimals = 2 },
    NOK = { code = "NOK", name = "Norwegian Krone", symbol = "kr", decimals = 2 },
    DKK = { code = "DKK", name = "Danish Krone", symbol = "kr", decimals = 2 },
    PLN = { code = "PLN", name = "Polish Zloty", symbol = "zł", decimals = 2 },
    RUB = { code = "RUB", name = "Russian Ruble", symbol = "₽", decimals = 2 },
    ZAR = { code = "ZAR", name = "South African Rand", symbol = "R", decimals = 2 },
    TRY = { code = "TRY", name = "Turkish Lira", symbol = "₺", decimals = 2 },
    THB = { code = "THB", name = "Thai Baht", symbol = "฿", decimals = 2 },
    MYR = { code = "MYR", name = "Malaysian Ringgit", symbol = "RM", decimals = 2 },
    IDR = { code = "IDR", name = "Indonesian Rupiah", symbol = "Rp", decimals = 2 },
    PHP = { code = "PHP", name = "Philippine Peso", symbol = "₱", decimals = 2 },
    VND = { code = "VND", name = "Vietnamese Dong", symbol = "₫", decimals = 0 },
    AED = { code = "AED", name = "UAE Dirham", symbol = "د.إ", decimals = 2 },
    SAR = { code = "SAR", name = "Saudi Riyal", symbol = "﷼", decimals = 2 },
    ILS = { code = "ILS", name = "Israeli Shekel", symbol = "₪", decimals = 2 },
    CZK = { code = "CZK", name = "Czech Koruna", symbol = "Kč", decimals = 2 },
    HUF = { code = "HUF", name = "Hungarian Forint", symbol = "Ft", decimals = 2 },
    RON = { code = "RON", name = "Romanian Leu", symbol = "lei", decimals = 2 },
    BGN = { code = "BGN", name = "Bulgarian Lev", symbol = "лв", decimals = 2 },
    HRK = { code = "HRK", name = "Croatian Kuna", symbol = "kn", decimals = 2 },
    ISK = { code = "ISK", name = "Icelandic Krona", symbol = "kr", decimals = 0 },
    CLP = { code = "CLP", name = "Chilean Peso", symbol = "$", decimals = 0 },
    COP = { code = "COP", name = "Colombian Peso", symbol = "$", decimals = 2 },
    PEN = { code = "PEN", name = "Peruvian Sol", symbol = "S/", decimals = 2 },
    ARS = { code = "ARS", name = "Argentine Peso", symbol = "$", decimals = 2 },
    BTC = { code = "BTC", name = "Bitcoin", symbol = "₿", decimals = 8 },
    ETH = { code = "ETH", name = "Ethereum", symbol = "Ξ", decimals = 8 },
}

-- Money metatable for OOP behavior.
local Money = {}
Money.__index = Money

--- Create a new Money object from minor units.
--- @param minor_units number Amount in minor units (cents, satoshis, etc.)
--- @param currency_code string ISO 4217 currency code
--- @return table|nil Money object or nil on error
--- @return string|nil Error message if creation failed
function Money.new(minor_units, currency_code)
    if type(minor_units) ~= "number" then
        return nil, "minor_units must be a number"
    end
    if type(currency_code) ~= "string" then
        return nil, "currency_code must be a string"
    end

    local currency_info = safe_currency.currency_codes[currency_code:upper()]
    if not currency_info then
        return nil, "Unknown currency code: " .. currency_code
    end

    local self = setmetatable({}, Money)
    self._minor_units = math.floor(minor_units)
    self._currency = currency_info
    return self
end

--- Create a Money object from major units.
--- @param major_units number Amount in major units (dollars, euros, etc.)
--- @param currency_code string ISO 4217 currency code
--- @return table|nil Money object or nil on error
--- @return string|nil Error message if creation failed
function Money.from_major(major_units, currency_code)
    if type(major_units) ~= "number" then
        return nil, "major_units must be a number"
    end

    local currency_info = safe_currency.currency_codes[currency_code:upper()]
    if not currency_info then
        return nil, "Unknown currency code: " .. currency_code
    end

    local multiplier = 10 ^ currency_info.decimals
    local minor_units = math.floor(major_units * multiplier + 0.5)

    return Money.new(minor_units, currency_code)
end

--- Create a zero Money object.
--- @param currency_code string ISO 4217 currency code
--- @return table|nil Money object or nil on error
--- @return string|nil Error message if creation failed
function Money.zero(currency_code)
    return Money.new(0, currency_code)
end

--- Get the currency code.
--- @return string
function Money:currency_code()
    return self._currency.code
end

--- Get the currency info.
--- @return table
function Money:currency()
    return self._currency
end

--- Get minor units (cents, satoshis, etc.).
--- @return number
function Money:minor()
    return self._minor_units
end

--- Get major units (truncated).
--- @return number
function Money:major()
    local divisor = 10 ^ self._currency.decimals
    return math.floor(self._minor_units / divisor)
end

--- Add two monetary values.
--- @param other table Money object to add
--- @return table|nil Money object or nil on error
--- @return string|nil Error message if operation failed
function Money:add(other)
    if getmetatable(other) ~= Money then
        return nil, "Cannot add non-Money value"
    end
    if self._currency.code ~= other._currency.code then
        return nil, "Currency mismatch: " .. self._currency.code .. " vs " .. other._currency.code
    end
    return Money.new(self._minor_units + other._minor_units, self._currency.code)
end

--- Subtract a monetary value.
--- @param other table Money object to subtract
--- @return table|nil Money object or nil on error
--- @return string|nil Error message if operation failed
function Money:sub(other)
    if getmetatable(other) ~= Money then
        return nil, "Cannot subtract non-Money value"
    end
    if self._currency.code ~= other._currency.code then
        return nil, "Currency mismatch: " .. self._currency.code .. " vs " .. other._currency.code
    end
    return Money.new(self._minor_units - other._minor_units, self._currency.code)
end

--- Multiply by a scalar.
--- @param scalar number The multiplier
--- @return table|nil Money object or nil on error
--- @return string|nil Error message if operation failed
function Money:mul(scalar)
    if type(scalar) ~= "number" then
        return nil, "Scalar must be a number"
    end
    return Money.new(math.floor(self._minor_units * scalar), self._currency.code)
end

--- Divide by a scalar.
--- @param scalar number The divisor
--- @return table|nil Money object or nil on error
--- @return string|nil Error message if operation failed
function Money:div(scalar)
    if type(scalar) ~= "number" then
        return nil, "Scalar must be a number"
    end
    if scalar == 0 then
        return nil, "Division by zero"
    end
    return Money.new(math.floor(self._minor_units / scalar), self._currency.code)
end

--- Check if zero.
--- @return boolean
function Money:is_zero()
    return self._minor_units == 0
end

--- Check if positive.
--- @return boolean
function Money:is_positive()
    return self._minor_units > 0
end

--- Check if negative.
--- @return boolean
function Money:is_negative()
    return self._minor_units < 0
end

--- Get absolute value.
--- @return table Money object
function Money:abs()
    return Money.new(math.abs(self._minor_units), self._currency.code)
end

--- Format as string.
--- @return string
function Money:format()
    local decimals = self._currency.decimals
    local divisor = 10 ^ decimals
    local abs_units = math.abs(self._minor_units)
    local major_part = math.floor(abs_units / divisor)
    local minor_part = abs_units % divisor
    local sign = self._minor_units < 0 and "-" or ""

    if decimals == 0 then
        return sign .. self._currency.symbol .. major_part
    else
        return string.format("%s%s%d.%0" .. decimals .. "d",
            sign, self._currency.symbol, major_part, minor_part)
    end
end

--- String representation.
function Money:__tostring()
    return self:format()
end

--- Equality comparison.
function Money:__eq(other)
    if getmetatable(other) ~= Money then
        return false
    end
    return self._minor_units == other._minor_units and
           self._currency.code == other._currency.code
end

--- Less than comparison (same currency only).
function Money:__lt(other)
    if getmetatable(other) ~= Money then
        return false
    end
    if self._currency.code ~= other._currency.code then
        return false
    end
    return self._minor_units < other._minor_units
end

--- Less than or equal comparison (same currency only).
function Money:__le(other)
    if getmetatable(other) ~= Money then
        return false
    end
    if self._currency.code ~= other._currency.code then
        return false
    end
    return self._minor_units <= other._minor_units
end

-- Expose the Money class
safe_currency.Money = Money

--- Parse currency code from string.
--- @param code_string string Currency code string
--- @return table|nil Currency info or nil if invalid
--- @return string|nil Error message if parsing failed
function safe_currency.parse_code(code_string)
    if type(code_string) ~= "string" then
        return nil, "Currency code must be a string"
    end
    local currency_info = safe_currency.currency_codes[code_string:upper()]
    if not currency_info then
        return nil, "Unknown currency code: " .. code_string
    end
    return currency_info
end

--- Check if valid currency code.
--- @param code_string string Currency code string
--- @return boolean
function safe_currency.is_valid_code(code_string)
    if type(code_string) ~= "string" then
        return false
    end
    return safe_currency.currency_codes[code_string:upper()] ~= nil
end

--- Get all available currency codes.
--- @return table Array of currency code strings
function safe_currency.available_codes()
    local codes = {}
    for code, _ in pairs(safe_currency.currency_codes) do
        table.insert(codes, code)
    end
    table.sort(codes)
    return codes
end

return safe_currency

File: bindings/scala/src/main/scala/proven/SafeCurrency.scala
// SPDX-License-Identifier: PMPL-1.0

package proven

import scala.util.Try

/**
 * ISO 4217 currency codes with metadata.
 */
enum CurrencyCode(
    val numericCode: Int,
    val minorUnits: Int,
    val name: String
):
  // Major currencies
  case USD extends CurrencyCode(840, 2, "US Dollar")
  case EUR extends CurrencyCode(978, 2, "Euro")
  case GBP extends CurrencyCode(826, 2, "Pound Sterling")
  case JPY extends CurrencyCode(392, 0, "Yen")
  case CHF extends CurrencyCode(756, 2, "Swiss Franc")
  case CAD extends CurrencyCode(124, 2, "Canadian Dollar")
  case AUD extends CurrencyCode(36, 2, "Australian Dollar")
  case NZD extends CurrencyCode(554, 2, "New Zealand Dollar")
  case CNY extends CurrencyCode(156, 2, "Yuan Renminbi")
  case HKD extends CurrencyCode(344, 2, "Hong Kong Dollar")
  case SGD extends CurrencyCode(702, 2, "Singapore Dollar")
  case SEK extends CurrencyCode(752, 2, "Swedish Krona")
  case NOK extends CurrencyCode(578, 2, "Norwegian Krone")
  case DKK extends CurrencyCode(208, 2, "Danish Krone")
  case KRW extends CurrencyCode(410, 0, "Won")
  case INR extends CurrencyCode(356, 2, "Indian Rupee")
  case RUB extends CurrencyCode(643, 2, "Russian Ruble")
  case BRL extends CurrencyCode(986, 2, "Brazilian Real")
  case ZAR extends CurrencyCode(710, 2, "Rand")
  case MXN extends CurrencyCode(484, 2, "Mexican Peso")
  case PLN extends CurrencyCode(985, 2, "Zloty")
  case TRY extends CurrencyCode(949, 2, "Turkish Lira")
  case THB extends CurrencyCode(764, 2, "Baht")
  case ILS extends CurrencyCode(376, 2, "New Israeli Sheqel")
  case AED extends CurrencyCode(784, 2, "UAE Dirham")
  case SAR extends CurrencyCode(682, 2, "Saudi Riyal")
  case CZK extends CurrencyCode(203, 2, "Czech Koruna")
  case HUF extends CurrencyCode(348, 2, "Forint")
  case PHP extends CurrencyCode(608, 2, "Philippine Peso")
  case TWD extends CurrencyCode(901, 2, "New Taiwan Dollar")
  case MYR extends CurrencyCode(458, 2, "Malaysian Ringgit")
  case IDR extends CurrencyCode(360, 2, "Rupiah")
  case VND extends CurrencyCode(704, 0, "Dong")
  case CLP extends CurrencyCode(152, 0, "Chilean Peso")
  case ARS extends CurrencyCode(32, 2, "Argentine Peso")
  case COP extends CurrencyCode(170, 2, "Colombian Peso")
  case PEN extends CurrencyCode(604, 2, "Sol")
  case EGP extends CurrencyCode(818, 2, "Egyptian Pound")
  case NGN extends CurrencyCode(566, 2, "Naira")
  case KES extends CurrencyCode(404, 2, "Kenyan Shilling")
  case GHS extends CurrencyCode(936, 2, "Ghana Cedi")
  case UAH extends CurrencyCode(980, 2, "Hryvnia")
  case RON extends CurrencyCode(946, 2, "Romanian Leu")
  case BGN extends CurrencyCode(975, 2, "Bulgarian Lev")
  case HRK extends CurrencyCode(191, 2, "Kuna")
  case ISK extends CurrencyCode(352, 0, "Iceland Krona")

  // Precious metals
  case XAU extends CurrencyCode(959, -1, "Gold")
  case XAG extends CurrencyCode(961, -1, "Silver")
  case XPT extends CurrencyCode(962, -1, "Platinum")
  case XPD extends CurrencyCode(964, -1, "Palladium")

  // Cryptocurrencies (non-ISO but commonly used)
  case BTC extends CurrencyCode(0, 8, "Bitcoin")
  case ETH extends CurrencyCode(0, 18, "Ether")
  case XRP extends CurrencyCode(0, 6, "XRP")
  case USDT extends CurrencyCode(0, 6, "Tether")
  case USDC extends CurrencyCode(0, 6, "USD Coin")

  /**
   * Get the multiplier for converting to minor units.
   */
  def minorUnitMultiplier: BigDecimal =
    if minorUnits < 0 then BigDecimal(1)
    else BigDecimal(10).pow(minorUnits)

  /**
   * Get the currency symbol (returns code if no symbol).
   */
  def symbol: String = this match
    case USD => "$"
    case EUR => "\u20AC"
    case GBP => "\u00A3"
    case JPY => "\u00A5"
    case CHF => "CHF"
    case CNY => "\u00A5"
    case INR => "\u20B9"
    case RUB => "\u20BD"
    case BRL => "R$"
    case KRW => "\u20A9"
    case TRY => "\u20BA"
    case THB => "\u0E3F"
    case ILS => "\u20AA"
    case PLN => "z\u0142"
    case BTC => "\u20BF"
    case ETH => "\u039E"
    case _ => toString

object CurrencyCode:
  private val byCode: Map[String, CurrencyCode] =
    CurrencyCode.values.map(c => c.toString -> c).toMap

  private val byNumeric: Map[Int, CurrencyCode] =
    CurrencyCode.values.filter(_.numericCode > 0).map(c => c.numericCode -> c).toMap

  /**
   * Parse a currency code from string (case-insensitive).
   */
  def parse(code: String): Option[CurrencyCode] =
    byCode.get(code.toUpperCase.trim)

  /**
   * Get currency by ISO numeric code.
   */
  def fromNumericCode(code: Int): Option[CurrencyCode] =
    byNumeric.get(code)

  /**
   * Check if a string is a valid currency code.
   */
  def isValid(code: String): Boolean =
    parse(code).isDefined

/**
 * Represents a monetary amount in a specific currency.
 * Internally stores the amount in minor units (e.g., cents for USD) to avoid floating-point issues.
 *
 * @param amountInMinorUnits The amount in the smallest currency unit
 * @param currency           The currency code
 */
case class Money private (
    amountInMinorUnits: Long,
    currency: CurrencyCode
):
  /**
   * Get the amount as a BigDecimal in major units.
   */
  def amount: BigDecimal =
    if currency.minorUnits <= 0 then BigDecimal(amountInMinorUnits)
    else BigDecimal(amountInMinorUnits) / currency.minorUnitMultiplier

  /**
   * Get the amount formatted with currency symbol.
   */
  def format: String =
    val formatted = formatAmount
    s"${currency.symbol}$formatted"

  /**
   * Get the amount formatted with currency code suffix.
   */
  def formatWithCode: String =
    val formatted = formatAmount
    s"$formatted ${currency.toString}"

  /**
   * Get the amount formatted without currency indicator.
   */
  def formatAmount: String =
    if currency.minorUnits <= 0 then
      amountInMinorUnits.toString
    else
      val scale = currency.minorUnits
      val major = amountInMinorUnits / currency.minorUnitMultiplier.toLong
      val minor = (amountInMinorUnits % currency.minorUnitMultiplier.toLong).abs
      f"$major%d.${minor.toString.reverse.padTo(scale, '0').reverse}"

  /**
   * Add two Money values (must be same currency).
   */
  def +(other: Money): Option[Money] =
    if currency != other.currency then None
    else
      Try(Math.addExact(amountInMinorUnits, other.amountInMinorUnits))
        .toOption
        .map(Money(_, currency))

  /**
   * Subtract two Money values (must be same currency).
   */
  def -(other: Money): Option[Money] =
    if currency != other.currency then None
    else
      Try(Math.subtractExact(amountInMinorUnits, other.amountInMinorUnits))
        .toOption
        .map(Money(_, currency))

  /**
   * Multiply by a scalar.
   */
  def *(scalar: Long): Option[Money] =
    Try(Math.multiplyExact(amountInMinorUnits, scalar))
      .toOption
      .map(Money(_, currency))

  /**
   * Multiply by a decimal scalar (rounds to nearest minor unit).
   */
  def *(scalar: BigDecimal): Option[Money] =
    val result = BigDecimal(amountInMinorUnits) * scalar
    if result > Long.MaxValue || result < Long.MinValue then None
    else Some(Money(result.setScale(0, BigDecimal.RoundingMode.HALF_EVEN).toLong, currency))

  /**
   * Divide by a scalar.
   */
  def /(divisor: Long): Option[Money] =
    if divisor == 0 then None
    else Some(Money(amountInMinorUnits / divisor, currency))

  /**
   * Negate the amount.
   */
  def unary_- : Option[Money] =
    Try(Math.negateExact(amountInMinorUnits))
      .toOption
      .map(Money(_, currency))

  /**
   * Get the absolute value.
   */
  def abs: Money =
    Money(Math.abs(amountInMinorUnits), currency)

  /**
   * Check if zero.
   */
  def isZero: Boolean = amountInMinorUnits == 0

  /**
   * Check if positive.
   */
  def isPositive: Boolean = amountInMinorUnits > 0

  /**
   * Check if negative.
   */
  def isNegative: Boolean = amountInMinorUnits < 0

  /**
   * Compare with another Money (must be same currency).
   */
  def compare(other: Money): Option[Int] =
    if currency != other.currency then None
    else Some(amountInMinorUnits.compare(other.amountInMinorUnits))

  override def toString: String = format

object Money:
  /**
   * Create Money from a BigDecimal amount in major units.
   */
  def apply(amount: BigDecimal, currency: CurrencyCode): Option[Money] =
    val inMinorUnits = amount * currency.minorUnitMultiplier
    if inMinorUnits > Long.MaxValue || inMinorUnits < Long.MinValue then None
    else Some(new Money(inMinorUnits.setScale(0, BigDecimal.RoundingMode.HALF_EVEN).toLong, currency))

  /**
   * Create Money from a double amount in major units.
   */
  def apply(amount: Double, currency: CurrencyCode): Option[Money] =
    apply(BigDecimal(amount), currency)

  /**
   * Create Money from minor units directly.
   */
  def fromMinorUnits(minorUnits: Long, currency: CurrencyCode): Money =
    new Money(minorUnits, currency)

  /**
   * Create zero Money in the given currency.
   */
  def zero(currency: CurrencyCode): Money =
    new Money(0, currency)

  /**
   * Parse a money string like "$123.45" or "123.45 USD".
   */
  def parse(input: String): Option[Money] =
    val trimmed = input.trim

    // Try "123.45 USD" format
    val withCodePattern = """^(-?[\d,.]+)\s*([A-Za-z]{3,4})$""".r
    trimmed match
      case withCodePattern(amountStr, codeStr) =>
        for
          currency <- CurrencyCode.parse(codeStr)
          amount <- Try(BigDecimal(amountStr.replace(",", ""))).toOption
          money <- Money(amount, currency)
        yield money
      case _ =>
        // Try "$123.45" format with common symbols
        val symbolMap = Map(
          "$" -> CurrencyCode.USD,
          "\u00A3" -> CurrencyCode.GBP,
          "\u20AC" -> CurrencyCode.EUR,
          "\u00A5" -> CurrencyCode.JPY,
          "\u20B9" -> CurrencyCode.INR
        )

        symbolMap.collectFirst {
          case (symbol, currency) if trimmed.startsWith(symbol) =>
            val amountStr = trimmed.stripPrefix(symbol).trim.replace(",", "")
            for
              amount <- Try(BigDecimal(amountStr)).toOption
              money <- Money(amount, currency)
            yield money
        }.flatten

/**
 * Safe currency operations with validation.
 */
object SafeCurrency:

  /**
   * Check if a string is a valid ISO 4217 currency code.
   */
  def isValidCode(code: String): Boolean =
    CurrencyCode.isValid(code)

  /**
   * Parse a currency code.
   */
  def parseCode(code: String): Option[CurrencyCode] =
    CurrencyCode.parse(code)

  /**
   * Get currency by numeric code.
   */
  def fromNumericCode(code: Int): Option[CurrencyCode] =
    CurrencyCode.fromNumericCode(code)

  /**
   * Get the number of minor units (decimal places) for a currency.
   */
  def getMinorUnits(code: String): Option[Int] =
    CurrencyCode.parse(code).map(_.minorUnits)

  /**
   * Get the currency symbol.
   */
  def getSymbol(code: String): Option[String] =
    CurrencyCode.parse(code).map(_.symbol)

  /**
   * Get the full currency name.
   */
  def getName(code: String): Option[String] =
    CurrencyCode.parse(code).map(_.name)

  /**
   * Create a Money value.
   */
  def money(amount: BigDecimal, currency: CurrencyCode): Option[Money] =
    Money(amount, currency)

  /**
   * Create a Money value from string code.
   */
  def money(amount: BigDecimal, currencyCode: String): Option[Money] =
    CurrencyCode.parse(currencyCode).flatMap(Money(amount, _))

  /**
   * Parse a money string.
   */
  def parseMoney(input: String): Option[Money] =
    Money.parse(input)

  /**
   * Format a monetary amount with the given currency.
   */
  def format(amount: BigDecimal, currency: CurrencyCode): String =
    Money(amount, currency).map(_.format).getOrElse(s"$amount ${currency.toString}")

  /**
   * Format a monetary amount with the given currency code.
   */
  def format(amount: BigDecimal, currencyCode: String): Option[String] =
    CurrencyCode.parse(currencyCode).flatMap(c => Money(amount, c).map(_.format))

  /**
   * Add two monetary amounts (must be same currency).
   */
  def add(a: Money, b: Money): Option[Money] = a + b

  /**
   * Subtract two monetary amounts (must be same currency).
   */
  def subtract(a: Money, b: Money): Option[Money] = a - b

  /**
   * Check if two currencies are the same.
   */
  def sameCurrency(a: Money, b: Money): Boolean =
    a.currency == b.currency

  /**
   * Convert amount from one currency to another using an exchange rate.
   * Rate should be: 1 unit of fromCurrency = rate units of toCurrency.
   */
  def convert(
      amount: Money,
      toCurrency: CurrencyCode,
      exchangeRate: BigDecimal
  ): Option[Money] =
    val converted = amount.amount * exchangeRate
    Money(converted, toCurrency)

  /**
   * Get all available currency codes.
   */
  def allCurrencyCodes: Seq[CurrencyCode] =
    CurrencyCode.values.toSeq

  /**
   * Get all fiat currency codes (exclude crypto and precious metals).
   */
  def fiatCurrencyCodes: Seq[CurrencyCode] =
    CurrencyCode.values.filter { c =>
      c.numericCode > 0 && c.numericCode < 900
    }.toSeq

  /**
   * Get all cryptocurrency codes.
   */
  def cryptoCurrencyCodes: Seq[CurrencyCode] =
    Seq(CurrencyCode.BTC, CurrencyCode.ETH, CurrencyCode.XRP, CurrencyCode.USDT, CurrencyCode.USDC)

  /**
   * Get all precious metal codes.
   */
  def preciousMetalCodes: Seq[CurrencyCode] =
    Seq(CurrencyCode.XAU, CurrencyCode.XAG, CurrencyCode.XPT, CurrencyCode.XPD)
